test/unit-test/CMakeLists.txt

```diff
@@ -30,11 +38,126 @@ set( KERNEL_DIRECTORY ${MODULE_ROOT_DIR}/test/FreeRTOS-Kernel )
     # Inform the user of the actions
-    message( STATUS "FreeRTOS-Kernel is required for this build. Submoduling it..." )
+    message( WARNING "FreeRTOS-Kernel is required for this build. Submoduling it..." )
     execute_process( COMMAND git submodule update --init --checkout ${KERNEL_DIRECTORY}
                      WORKING_DIRECTORY ${MODULE_ROOT_DIR} )
 endif()
 
+# Define a CMock resource path.
+set( CMOCK_DIR ${MODULE_ROOT_DIR}/tools/CMock CACHE INTERNAL "CMock library source directory." )
+
+# Check if the CMock source directory exists, and if not present, clone the submodule
+# if BUILD_CLONE_SUBMODULES configuration is enabled.
+if( NOT EXISTS ${CMOCK_DIR}/src )
+    # Attempt to clone CMock and inform the user of the actions.
+    message( WARNING "CMock is required for this build. Submoduling it..." )
+    execute_process( COMMAND git submodule update --init --checkout --recursive ${CMOCK_DIR}
+                     WORKING_DIRECTORY ${MODULE_ROOT_DIR} )
+endif()
+
+
+if( NOT EXISTS ${CMOCK_DIR}/vendor/unity/docs )
+    # Attempt to clone Unity and inform the user of the actions.
+    message( WARNING "Unity is required for this build. Submoduling it..." )
+    execute_process( COMMAND git submodule update --init --checkout --recursive ${CMOCK_DIR}
+                     WORKING_DIRECTORY ${MODULE_ROOT_DIR} )
+endif()
+
+# Check for the platform/OS specific details only if user has allowed the Cmake to
+# install tools.
+if( ${ALLOW_CMAKE_TO_INSTALL_TOOLS} )
+    # If the platform is Ubuntu.
+    if( ${UT_PLATFORM} STREQUAL "Ubuntu" )
+        set( SYSTEM_PKG_MNGR "apt-get" )
+        set( SYSTEM_INSTALL_CMD "install" )
+    # Or the platform is MacOS.
+    elseif( ${UT_PLATFORM} STREQUAL "MacOS" )
+        set( SYSTEM_PKG_MNGR "brew" )
+        set( SYSTEM_INSTALL_CMD "install" )
+    # Or the platform is rpm based system.
+    elseif( ${UT_PLATFORM} STREQUAL "RPMSystem" )
+        set( SYSTEM_PKG_MNGR "yum" )
+        set( SYSTEM_INSTALL_CMD "install" )
+    else()
+        # Otherwise, check whether user has given any system package manager and
+        # corresponding command.
+        if( ( "${SYSTEM_PKG_MNGR}" STREQUAL "" ) OR ( "${SYSTEM_INSTALL_CMD}" STREQUAL "" ) )
+            message( FATAL_ERROR "Wrong platform specified. While invoking the cmake command, you can choose the platform as shown below\n"
+                                 "Only 'Ubuntu', 'RPMSystem' and 'MacOS' are valid options.\n"
+                                 "    cmake -S. -B build <Additional commands/options> -DUT_PLATFORM=Ubuntu\n"
+                                 "Otherwise, you can set the package manager and the install command. Example for Red Hat Enterprise Linux:\n"
+                                 "    cmake -S. -B build <Additional commands/options> -DSYSTEM_PKG_MNGR=yum -DSYSTEM_INSTALL_CMD=install" )
+        endif()
+    endif()
+endif()
+
+# Find 'unifdef' utility.
+find_program( UNIFDEF_found unifdef )
+
+if( NOT UNIFDEF_found )
```

Comment:
This seems like a repeated pattern for checking and installing tool. Should we write a function?

Reply:
Yes, I agree. I made the change in the commit https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/404/commits/b16fe1172e251f02eb0ba9655f48987f346f84be